### PR TITLE
auto-update: validate container image

### PR DESCRIPTION
**Pull request by vrothberg**
Auto updates require containers to be created with a fully-qualified image reference.  Short names are not supported due the ambiguity of their source registry.  Initially, container creation errored out for non FQN images but it seems that Podman has regressed.

Fixes: #15879
Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
Auto update: error out on container creation for non-FQN images.
```

### CODE REVIEW

**Comment by vrothberg** [comment]
@rhatdan @mheon PTAL

File: test/system/255-auto-update.bats [hunk 1]
```diff
@@ -133,6 +133,25 @@ function _confirm_update() {
...
+    # Fully-qualified image reference is requred
+    run_podman create --label io.containers.autoupdate=registry $IMAGE
+    run_podman rm -f "$output"
+
+    # Short name does not work
+    shortname="shortname:latest"
+    run_podman image tag $IMAGE shortname:latest
+    run_podman 125 create --label io.containers.autoupdate=registry $shortname
+    is "$output" "Error: short name: auto updates require fully-qualified image reference: \"$shortname\""
+#    run_podman untag $shortname
```
Comment (edsantiago) [inline test/system/255-auto-update.bats:146]: Detritus

File: test/system/255-auto-update.bats [hunk 2]
```diff
@@ -133,6 +133,25 @@ function _confirm_update() {
     die "Timed out waiting for $cname to update; old IID=$old_iid"
 }
 
+@test "podman auto-update - validate inpute" {
+    # Fully-qualified image reference is requred
```
Comment (edsantiago) [inline test/system/255-auto-update.bats:137]: Typo, 'required'. (Not worth mentioning by itself, but line 146 is).

File: test/system/255-auto-update.bats [hunk 3]
```diff
@@ -133,6 +133,25 @@ function _confirm_update() {
     die "Timed out waiting for $cname to update; old IID=$old_iid"
 }
 
+@test "podman auto-update - validate inpute" {
```
Comment (edsantiago) [inline test/system/255-auto-update.bats:136]: typo, "input"

File: test/system/255-auto-update.bats [hunk 4]
```diff
@@ -133,6 +133,25 @@ function _confirm_update() {
...
+    # Fully-qualified image reference is requred
+    run_podman create --label io.containers.autoupdate=registry $IMAGE
+    run_podman rm -f "$output"
+
+    # Short name does not work
+    shortname="shortname:latest"
+    run_podman image tag $IMAGE shortname:latest
```
Comment (edsantiago) [inline test/system/255-auto-update.bats:143]: Please use your shiny new variable defined in the line above

File: test/system/255-auto-update.bats [hunk 5]
```diff
@@ -133,6 +133,25 @@ function _confirm_update() {
...
+    run_podman 125 create --label io.containers.autoupdate=registry $shortname
+    is "$output" "Error: short name: auto updates require fully-qualified image reference: \"$shortname\""
+#    run_podman untag $shortname
+
+    # Requires docker (or no) transport
+    archive=$PODMAN_TMPDIR/archive.tar
+    run_podman save -o $archive $IMAGE
+    run_podman 125 create --label io.containers.autoupdate=registry docker-archive:$archive
+    is "$output" ".*Error: auto updates require the docker image transport but image is of transport \"docker-archive\""
+}
```
Comment (edsantiago) [inline test/system/255-auto-update.bats:153]: Please clean up: `run_podman rmi $shortname`

**Review by edsantiago — CHANGES_REQUESTED** [review]
Minor nits adding up.

Also... the first half of the test is actually applicable to podman-remote, but it is not run (all tests in this bats file are skipped on remote). Would there be any point to moving this test to another bats file? (My inclination is no, but I think it's worth an explicit decision)

**Comment by vrothberg** [comment]
Thanks for the review, @edsantiago!  Wow, how much a phone call can distract me.

**Comment by vrothberg** [comment]
> Would there be any point to moving this test to another bats file? (My inclination is no, but I think it's worth an explicit decision)

I do not feel strongly either way.  If you have a preference, I happily move the test around.

**Comment by edsantiago** [comment]
Is there any conceivable situation in which this could happen?
```console
# podman create --label ... shortname:latest
Error: you can't do that
# podman-remote !*
[runs just fine]
```
I'm guessing no. But if there is any possible code flow where that could pass, then it might be worth adding a regression test.

**Comment by vrothberg** [comment]
> Is there any conceivable situation in which this could happen?

No, there is none to my knowledge.  There are various ways to create a container but all will be validated.  That's why I moved it there.  Initially, the check was in specgen which clearly broke at some point (also due to the missing tests).

**Comment by rhatdan** [comment]
LGTM

**Comment by vrothberg** [comment]
> LGTM. Tests are red.

Fixed.  There was a redundant test that I removed.  @edsantiago, can you give your blessing for that?

**Comment by edsantiago** [comment]
OK hold on: is it intentional to revert #10063 then? Because it looks like that was deliberate. And if that will be reverted, it looks like we need many more reversions, such as to the man pages and the code?

**Comment by vrothberg** [comment]
> OK hold on: is it intentional to revert #10063 then? Because it looks like that PR was deliberate. And if that will be reverted, it looks like we need many more reversions, such as to the man pages and the code?

That is a fair point.  It would break backwards compat as the "local" policy allows short names.  In retrospect, I think that was a bad idea but that doesn't justify breaking things.  Thank you for catching, Ed.

**Pull request merged**